### PR TITLE
 HPCC-31227 Address Prometheous histogram issues

### PR DESCRIPTION
**Pull request by rpastrana**
- Fixes invalid quote on le=+Inf bucket
 - Qualifies sum and count w/ metric name
 - Normalizes all names to lowercase per promtool suggestion

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by rpastrana** [comment]
@ghalliday targeting latest point branch, but it seems this feature has been broken since its inception (https://github.com/hpcc-systems/HPCC-Platform/pull/16624) which seems to have been included as far back as 8.12.x

File: system/metrics/sinks/prometheus/prometheusSink.cpp [hunk 1]
```diff
@@ -180,6 +180,9 @@ void PrometheusMetricSink::toPrometheusHistogram(const std::string &name, const
 std::string PrometheusMetricSink::getPrometheusMetricName(const std::shared_ptr<IMetric> &pMetric)
 {
     std::string name = pMetric->queryName();
+    // Convert to lower case - per promtool output: "metric names should be written in 'snake_case' not 'camelCase'"
```
Comment (rpastrana) [inline system/metrics/sinks/prometheus/prometheusSink.cpp:183]: Perhaps out of scope of this issue, but lower casing all metric names per Prometheus promtool output:
dfs_8880_WsDfs_KeepAlive_seconds metric names should be written in 'snake_case' not 'camelCase'
dfs_8880_WsDfs_KeepAlive_seconds no help text

File: system/metrics/sinks/prometheus/prometheusSink.cpp [hunk 2]
```diff
@@ -240,16 +243,16 @@ const std::vector<std::string> &PrometheusMetricSink::getHistogramLabels(const s
 
     // Add the inf label
     std::string infLabel(name);
-    infLabel.append("_bucket{\"le=+Inf\"}");
+    infLabel.append("_bucket{le=\"+Inf\"}");
```
Comment (rpastrana) [inline system/metrics/sinks/prometheus/prometheusSink.cpp:246]: Incorrect double quote location

File: system/metrics/sinks/prometheus/prometheusSink.cpp [hunk 3]
```diff
@@ -240,16 +243,16 @@ const std::vector<std::string> &PrometheusMetricSink::getHistogramLabels(const s
...
     labels.emplace_back(infLabel);
 
     // Sum and count
-    std::string sumLabel;
-    sumLabel.append("_sum");
+    std::string sumLabel(name);
```
Comment (rpastrana) [inline system/metrics/sinks/prometheus/prometheusSink.cpp:250]: _sum and _count missing metric name prefix

**Comment by rpastrana** [comment]
No more "INVALID" errors in prometheous ui:
image

Sample histogram metric accepted by prom chart ui:
image

File: system/metrics/sinks/prometheus/prometheusSink.cpp [hunk 4]
```diff
@@ -240,16 +243,16 @@ const std::vector<std::string> &PrometheusMetricSink::getHistogramLabels(const s
...
+    sumLabel.append("_sum{}");
     labels.emplace_back(sumLabel);
 
-    std::string countLabel;
-    countLabel.append("_count");
+    std::string countLabel(name);
+    countLabel.append("_count{}");
```
Comment (kenrowland) [inline system/metrics/sinks/prometheus/prometheusSink.cpp:255]: Curious why the '{}" is present. Braces are not part of what is shown on the Prometheus page documenting their exposition formats.  See https://prometheus.io/docs/instrumenting/exposition_formats/

Please double check the reason for "{}"

**Review by kenrowland — CHANGES_REQUESTED** [review]

File: system/metrics/sinks/prometheus/prometheusSink.cpp (hunk 4 above)
Comment (rpastrana) [reply system/metrics/sinks/prometheus/prometheusSink.cpp:255]: Great observation, it's legal syntax as confirmed by promtool and prometheous graph UI used as a label container. It's optional, and honestly it was added as I tested why your histograms were deemed invalid.

Comment (rpastrana) [reply system/metrics/sinks/prometheus/prometheusSink.cpp:255]: What stands out to me is the fact there's currently no mechanism (at least not one I could see) to populate the histogram buckets w/ any available metadata as labels. That will likely need to be fixed.

**Review by kenrowland — APPROVED** [review]
Based on comment that promtool accepted the {} appended to the histogram sum and count outputs.

**Pull request merged**